Woodoku/src/Entity/WoodokuBoard.py

```diff
@@ -25,89 +28,171 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
+        print(all_blocks)
+
+        for row, col in all_blocks:
+            if self.can_add_shape_at_location(shape, x=row, y=col):
+                return True
+
+        return False
 
     def can_add_shape_at_location(self, shape: WoodokuShape, x: int, y: int) -> bool:
+        """Check if the woodoku `shape` can be placed into the board at location
+        `(x, y)` coordinates as the left upper corner of the shape.
+
+        Args:
+            shape (WoodokuShape): The shape needed to be checked
+            x (int): x coordinate
+            y (int): y coordinate
+
+        Returns:
+            bool: if `shape` can be added to `(x,y)`
         """
-        Check if the woodoku shape can be placed into the board at location (x, y) coordinates as the left upper corner
-        of the shape.
-        :param shape: the shape
-        :param x: x coordinate
-        :param y: y coordinate
-        :return: if can be placed at location.
-        """
-        pass
+        blocks = self.__map_shape_to_board_at(shape, x, y)
+        return not self.__representation.is_occupied(blocks)
 
     def add_shape(self, shape: WoodokuShape, x: int, y: int) -> None:
-        """
-        Add the shape to woodoku at coordinate (x, y). This method should be called after checking the addibility. If
-        unable to add, raise exception.
+        """Add the shape to woodoku at coordinate (x, y). Only called if the shape
+        can be added at `(x,y)`.
 
         This method includes actions:
-        1. Clear winning row/column/3x3 block
+        1. Clear winning row/column/3x3 box
         2. Call scoreAgent to calculate score
 
-        :param shape: The shape
-        :param x: x coordinate
-        :param y: y coordinate
-        :return: none
+        Args:
+            shape (WoodokuShape): The shape to be added
+            x (int): x coordinate
+            y (int): y coordinate
         """
-        pass
+        # add shape to block
+        shape_blocks = shape.map_to_board_at(x, y)
+        self.__representation.add_blocks(shape_blocks)
+
+        # determine groups and clear the groups
+        group_info, group_blocks = self.__find_groups()
+        self.__scoreAgent.calculate_winning(group_info)
+        self.__representation.remove_blocks(group_blocks)
 
     def get_score(self) -> int:
-        pass
+        return self.__scoreAgent.get_score()
 
-    def __find_winning_blocks(self) -> Dict[str, int]:
-        """
-        Check current board and see if there is any winning blocks such as complete rows, columns or 3x3 block.
-        :return: A dictionary mapping from name of the winning block to the number of winning block.
-            e.g. { "row": 1, "column": 2, "3x3 block": 0 }
-        """
-        pass
+    def __find_groups(self) -> Tuple[Dict[str, int], Set[Tuple[int, int]]]:
+        """Check current board and see if there is any groups such as
+        complete rows, columns or 3x3 box and report them. Remove all groups.
```

Comment:
Seems that it is not removing all the groups. Might want to update the docstring.